### PR TITLE
When calling selectAll() attributes should be updated

### LINKED ISSUE #247: grid.selection.selectAll() doesn’t update the selection-mode attribute
If you call `grid.selection.selectAll()`, the `selection-mode` attribute on the element is not updated to “all”. The `grid.selection.mode` property returns “all” correctly.

At the same time,  check that setting the select.mode property reflects correctly, and that selection.clear() correctly sets the attribute and property from "all" to "multi" (when in single-select mode, the mode doesn’t change when clear() is called).

### PR DESCRIPTION
**Pull request by manolo**
[Review on Reviewable](https://reviewable.io/reviews/vaadin/vaadin-grid/264)

Fixes #247

### CODE REVIEW

**Comment by tomivirkki** [comment]
This branch should base directly on master, now it bases on a commit on 12. november and includes some unrelated commits. Please _rebase_

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

_[java/src/main/java/com/vaadin/elements/grid/selection/IndexBasedSelectionModel.java, line 20 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3_2ntwPgZ8GCdWHNYg:-K3_2ntxU0PwV8rD1XFD:-898347186) ([raw file](https://github.com/vaadin/vaadin-grid/blob/c8985964f6b5cd68f73a216ed31ab70542206183/java/src/main/java/com/vaadin/elements/grid/selection/IndexBasedSelectionModel.java#L20)):_
No need for a new event type. This approach is overly complex and introduces some regression as well.

All that is needed in the patch is 3 LOC:
1. event name fix in vaadin-grid.html
2. grid.fireEvent(new SelectAllEvent<Object>(this)); fired in multi-model selectAll method
3. Same event fired in multi-model reset()

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-:-K3_2Q868ZZsBpZ_tzX6:2038226318)_

**Comment by manolo** [comment]
Yep I based in on the 'review doc' branch, will rebase with master when that is merged

---

Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-:-K3_7-wvCcXenvpcjpmI:-688792719)_

**Comment by manolo** [comment]
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

_[java/src/main/java/com/vaadin/elements/grid/selection/IndexBasedSelectionModel.java, line 20 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3_2ntwPgZ8GCdWHNYg:-K3_7Iwn95RNLn7jRPfC:1414062832) ([raw file](https://github.com/vaadin/vaadin-grid/blob/c8985964f6b5cd68f73a216ed31ab70542206183/java/src/main/java/com/vaadin/elements/grid/selection/IndexBasedSelectionModel.java#L20)):_
I'm ok if we want to change the current design. 
The responsibility of changing the selection model, and doing things when selectAll is called is given to the SelectionModel implementation.
If you take a look to 'all' implementation, selectAll implies a change of selection model apart from other actions, in the case of 'single' the model does not changes  although it changes selections.

Summing up, if we do something simpler like selectAll is only allowed in 'multi' mode we can extract the logic to JS, and probably save a lot of code which I would  suggest to do in this story.  It implies of course to modify some use cases in tests like single -> selectAll -> selection is reset to empty.

Otherwise if the responsibility of changing modes, reseting selections, etc is in the selection mode in Java, we need a mechanism to communicate those changes to JS, for that is needed events.

So let me know what is the way you want and will do in that way. 

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264)_

**Comment by manolo** [comment]
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

_[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 373 \[r5\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3se9DzSS-BHuGkuHyj:-K3se9DzSS-BHuGkuHyk:-477494116) ([raw file](https://github.com/vaadin/vaadin-grid/blob/3ee997256c33a012f63c314d70b528afb940971a/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L373)):_
@tomivirkki IMO all this code here makes no so much sense if the selection logic is in the model.
I think all the problems come from the model-multi and model-all sharing the same class. Probably having two classes this code should be as simpler as `grid.setSelectionModel(mode.createModel());`. But it's something we could discuss later in a separate card.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264)_

**Comment by Saulis** [comment]
Review status: 0 of 8 files reviewed at latest revision, 14 unresolved discussions.

---

_[test/grid-selecting-rows.html, line 296 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3sddpiQQlFK4DVlsZu:-K3sddpiQQlFK4DVlsZv:-622179909) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L296)):_
This test is now testing two separate features, the attribute test should be separate.

---

_[test/grid-selecting-rows.html, line 304 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3sdxtCG3kxoy8btf9T:-K3sdxtCG3kxoy8btf9U:-1362196105) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L304)):_
We should extract these into 
````javascript
function clickSelectAll() {
  var inputs = qaLocal('.vaadin-grid-header input');
  triggerMouseEvent(inputs[0], "click");
}
````

---

_[test/grid-selecting-rows.html, line 309 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3sdvn9vx5P9G8r8UOG:-K3sdvn9vx5P9G8r8UOH:245111761) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L309)):_
This also needs a separate test

---

_[test/grid-selecting-rows.html, line 314 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3seenvPJgLXP7woMN_:-K3seenvPJgLXP7woMNa:2116055711) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L314)):_
This spy should be a global or suite variable (with a descriptive name, `onSelectionModeChangeListener` for example) and attached + reset in global `beforeEach` /  `afterEach`

---

_[test/grid-selecting-rows.html, line 487 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3sbw1y32vDN9nnp1n_:-K3sbw1y32vDN9nnp1na:1728715579) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L487)):_
Could you rephrase this to `should clear selection` please?

---

_[test/grid-selecting-rows.html, line 491 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3scDhYFy0Lh9EaJMdH:-K3scDhYFy0Lh9EaJMdI:-373477306) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L491)):_
Could you rephrase this to `should not change selection when selecting all` ?

---

_[test/grid-selecting-rows.html, line 557 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3scdzWPvCtshrsZDjG:-K3scdzWPvCtshrsZDjH:1123938425) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L557)):_
`should clear selection`

---

_[test/grid-selecting-rows.html, line 561 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3scTRdGwqGaVA8YEaD:-K3scTRdGwqGaVA8YEaE:446530339) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L561)):_
This should have a separate test `should change selection mode when clearing`

---

_[test/grid-selecting-rows.html, line 564 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3sciNKQjD-qa5HXiqo:-K3sciNKQjD-qa5HXiqp:-1370718380) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L564)):_
`should change selection mode when selecting all`

---

_[test/grid-selecting-rows.html, line 642 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3scusLokLm0EGiK_3p:-K3scusLokLm0EGiK_3q:190629296) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L642)):_
`should change selection mode when clearing`

---

_[test/grid-selecting-rows.html, line 648 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3scxLeFeB7MDRMkNAm:-K3scxLeFeB7MDRMkNAn:1524695921) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L648)):_
`should select all items`

---

_[test/grid-selecting-rows.html, line 651 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3sd7OUA_QcACDgqCJx:-K3sd7OUA_QcACDgqCJy:1226047303) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L651)):_
`should not change selection mode when selecting all`

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264)_

**Comment by tomivirkki** [comment]
_[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 373 \[r5\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3se9DzSS-BHuGkuHyj:-K3sfGdB9jHxZSiP_z_5:141959635) ([raw file](https://github.com/vaadin/vaadin-grid/blob/3ee997256c33a012f63c314d70b528afb940971a/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L373)):_
Well they used to be separate models before but it had some issues

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264)_

**Comment by manolo** [comment]
Review status: 0 of 8 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

_[test/grid-selecting-rows.html, line 309 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3sdvn9vx5P9G8r8UOG:-K3skUCHuQ4X3svk9C_O:1602522773) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L309)):_
I think the functionality is so trivial that it's not worth to increase the number of tests adding more time to the suite. I don't think  one-assertion === one-test should be always true. But anymore this card had a '2' estimation, probably we should consider all this boiler plate each time we estimate apparently simple tasks.

---

_[test/grid-selecting-rows.html, line 314 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3seenvPJgLXP7woMN_:-K3slF-zHYmU-qPTGXCM:-1887868892) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L314)):_
They are different cases. in the first case spy is added at the beginning but in the second case after we have interacted with the grid (first click)

---

_[test/grid-selecting-rows.html, line 487 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3sbw1y32vDN9nnp1n_:-K3slXOh8wKESKqz7SRx:435469276) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L487)):_
it was a previous comment

---

_[test/grid-selecting-rows.html, line 651 \[r4\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3sd7OUA_QcACDgqCJx:-K3slo6cpHRxZ7Fiofbt:940224398) ([raw file](https://github.com/vaadin/vaadin-grid/blob/82548cbd3a0edabaa64b4c1f7ac612c43cdb1ad3/test/grid-selecting-rows.html#L651)):_
All this rewording could be done in a separate PR, it's unrelated to this story I think

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264)_

**Comment by manolo** [comment]
Review status: 0 of 8 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

_[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 373 \[r5\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3se9DzSS-BHuGkuHyj:-K3smfm0cRuRbE-wOMDb:803643122) ([raw file](https://github.com/vaadin/vaadin-grid/blob/3ee997256c33a012f63c314d70b528afb940971a/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L373)):_
Which issues? now also it has issues. If we have the abstraction of selection-models to isolate and sinplify behaviour implementations, I don't think we should move any logic out of those classes, otherwise the designed pattern is incorrect.

---

_[java/src/main/java/com/vaadin/elements/grid/selection/IndexBasedSelectionModel.java, line 20 \[r2\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3_2ntwPgZ8GCdWHNYg:-K3sm-6sXMrKaxIfJR2S:-499146009) ([raw file](https://github.com/vaadin/vaadin-grid/blob/c8985964f6b5cd68f73a216ed31ab70542206183/java/src/main/java/com/vaadin/elements/grid/selection/IndexBasedSelectionModel.java#L20)):_
Done.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264)_

**Comment by tomivirkki** [comment]
_[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 373 \[r5\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3se9DzSS-BHuGkuHyj:-K3t8GPg6GM9fwq_Cdkn:1599275744) ([raw file](https://github.com/vaadin/vaadin-grid/blob/3ee997256c33a012f63c314d70b528afb940971a/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L373)):_
Unlike Grid itself, we switch selection mode when toggling the "select-all-checkbox". Changing the model internally in Grid forces lots of size-recalculations etc which cause unwanted flickering. That was one of the reasons for the change.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264)_

**Comment by manolo** [comment]
Review status: 0 of 8 files reviewed at latest revision, 14 unresolved discussions.

---

*[java/src/main/java/com/vaadin/elements/grid/GridElement.java, line 373 \[r5\]](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264#-K3se9DzSS-BHuGkuHyj:-K3xP_Zn2GwgKNnSjD23:606136961) ([raw file](https://github.com/vaadin/vaadin-grid/blob/3ee997256c33a012f63c314d70b528afb940971a/java/src/main/java/com/vaadin/elements/grid/GridElement.java#L373)):*
Well, I think we could have separate classes for 'multi' and 'all', 'all' knows all 'multi' internals , then and because in setSelectionMode we call the abstract createModel() method of the model, in the 'all' implementation, it should have the responsibility of returning the current multi or a new all, so 'all' is a wrapper for 'multi' but all hacks and optimitations are in the 'all' model.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/vaadin/vaadin-grid/264)*
